x-pack/legacy/plugins/uptime/public/pages/settings.tsx

```diff
@@ -70,52 +67,32 @@ export const SettingsPage = () => {
     dispatch(getDynamicSettings());
   }, [dispatch]);
 
-  const [formFields, setFormFields] = useState<DynamicSettings | null>(dss.settings || null);
-
-  if (!dss.loadError && formFields == null && dss.settings) {
-    setFormFields({ ...dss.settings });
-  }
+  const [formFields, setFormFields] = useState<DynamicSettings>({ ...dss.settings });
 
   const fieldErrors = getFieldErrors(formFields);
 
   const isFormValid = !(fieldErrors && Object.values(fieldErrors).find(v => !!v));
 
-  const onChangeFormField: OnFieldChangeType = (field, value) => {
-    if (formFields) {
-      const newFormFields = cloneDeep(formFields);
-      set(newFormFields, field, value);
-      setFormFields(cloneDeep(newFormFields));
-    }
-  };
+  const onChangeFormField: OnFieldChangeType = changedField =>
+    setFormFields(Object.assign({ ...merge(formFields, changedField) }));
```

Comment:
We are using merge bcoz Object.assign doesn't take of nested merging, right?

Reply:
I've gotten rid of merge altogether because it isn't pure and it was editing the state var. I'm wondering if settings shouldn't just be a flat object to avoid recursive concerns when editing slices of it.

Reply:
yes i think having a right data structure is important, i do agree, having nested have made a lot of complexity. if you think, it's worth an effort, go ahead.

Reply:
I'll make a follow-up issue.

We have other more important stuff we need to review and it's not worth holding that up for more work in this patch. We don't have any further settings changes planned for the time being, so this complexity shouldn't increase any more than what it is right now until we can get to it. It should be a light patch.

Reply:
Issue: https://github.com/elastic/uptime/issues/175